### ISSUE TITLE
TypeScript example doesn't work

### ISSUE THREAD

**Issue by mxs42** [evanshortiss/express-joi-validation#18]
```
import express from "express";
import {ContainerTypes, createValidator, ValidatedRequest, ValidatedRequestSchema} from "express-joi-validation";

import * as joi from "@hapi/joi"

const route = express.Router();
const validator = createValidator();
const querySchema = joi.object({
  name: joi.string().required()
});

interface HelloRequestSchema extends ValidatedRequestSchema {
  [ContainerTypes.Query]: joi.extractType<typeof querySchema>
}

route.get(
  "/hello",
  validator.query(querySchema),
  (req: ValidatedRequest<HelloRequestSchema>, res) => {
    res.end(`Hello ${req.query.name}`)
  }
);
```

```
Error:(21, 3) TS2345: Argument of type '(req: ValidatedRequest<HelloRequestSchema>, res: Response) => void' is not assignable to parameter of type 'RequestHandlerParams'.
  Type '(req: ValidatedRequest<HelloRequestSchema>, res: Response) => void' is not assignable to type 'RequestHandler'.
    Types of parameters 'req' and 'req' are incompatible.
      Type 'Request' is missing the following properties from type 'ValidatedRequest<HelloRequestSchema>': fields, originalBody, originalQuery, originalHeaders, and 2 more.
```

**Comment by evanshortiss** [comment]
@mxs42 can you provide more details such as your `tsconfig.json` and `package.json`?

I compiled your code without issue using the following tsconfig and package. I also added the line `import "joi-extract-type"` after you imported joi.

`tsconfig.json`
```
{
    "files": ["src/index.ts"],
    "compilerOptions": {
        "esModuleInterop": true
    }
}
```

`package.json`
```
{
  "name": "ts-example-joi",
  "version": "1.0.0",
  "description": "",
  "main": "src/index.js",
  "dependencies": {
    "@hapi/joi": "~15.1.0",
    "express": "~4.17.1",
    "express-joi-validation": "~2.0.0",
    "joi-extract-type": "~15.0.0"
  },
  "devDependencies": {
    "@types/express": "~4.17.0",
    "@types/hapi__joi": "~15.0.2",
    "@types/node": "~12.6.8",
    "typescript": "~3.5.3"
  },
  "scripts": {
    "compile": "tsc",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "Evan Shortiss",
  "license": "MIT"
}
```

**Comment by evanshortiss** [comment]
@mxs42 closing this for now. Reopen it if you're still having issues.

**Comment by hijarian** [comment]
I got **exactly the same** error just now, with production (staging) code.

`package.json`:
```json
{
  "scripts": {
    "build": "yarn run build:src && yarn run migrate",
    "develop": "ts-node-dev --respawn --transpileOnly src/api/index.ts",
    "serve": "node ./dist/api/index.js",
    "build:src": "tsc && yarn run copy-static-assets",
    "test": "jest && stryker run",
    "test:coverage": "jest --coverage",
    "test:watch": "jest --watch",
    "test:mutations": "stryker run",
    "clean": "rimraf .stryker-tmp dist coverage reports",
    "migrate": "knex --migrations-directory ./src/migrations --knexfile dist/utils/database/knexfile.js  migrate:latest",
    "copy-static-assets": "ts-node copyStaticAssets.ts",
    "reset-database": "ts-node resetDatabase.ts"
  },
  "husky": {
    "hooks": {
      "pre-commit": "tsc --noEmit && lint-staged"
    }
  },
  "lint-staged": {
    "*.{js,ts,tsx,jsx,json,md}": [
      "eslint . --fix",
      "git add"
    ]
  },
  "dependencies": {
    "@hapi/joi": "^15.1.1",
    "@sentry/node": "^5.5.0",
    "animals": "^0.0.3",
    "body-parser": "^1.19.0",
    "connect-redis": "^3.4.1",
    "cookie-parser": "^1.4.4",
    "cors": "^2.8.5",
    "dotenv-flow": "^0.4.0",
    "express": "^4.17.1",
    "express-joi-validation": "^2.0.0",
    "express-session": "^1.16.2",
    "helmet": "^3.19.0",
    "isbot": "^2.3.2",
    "jsonwebtoken": "^8.5.1",
    "juice": "^5.2.0",
    "knex": "^0.19.0",
    "lodash": "4.17",
    "log4js": "^4.1.0",
    "mailchimp-api-v3": "^1.13.0",
    "moment": "^2.24.0",
    "morgan": "^1.9.1",
    "nodemailer": "^6.3.0",
    "nodemailer-mandrill-transport": "^1.2.0",
    "passport": "^0.4.0",
    "passport-facebook-token": "^3.3.0",
    "passport-jwt": "^4.0.0",
    "passport-local": "^1.0.0",
    "paypal-ipn": "^3.0.0",
    "paypal-rest-sdk": "^2.0.0-rc.2",
    "pg": "^7.12.0",
    "pug": "^2.0.4",
    "redis": "^2.8.0",
    "request": "^2.88.0",
    "request-promise-native": "^1.0.7",
    "stripe": "^7.4.0",
    "twilio": "^3.33.1",
    "joi-extract-type": "^15.0.0"
  },
  "devDependencies": {
    "@stryker-mutator/core": "^1.2.0",
    "@stryker-mutator/html-reporter": "^1.2.0",
    "@stryker-mutator/jest-runner": "^1.2.0",
    "@stryker-mutator/typescript": "^1.2.0",
    "@types/body-parser": "^1.17.0",
    "@types/connect-redis": "^0.0.10",
    "@types/cookie-parser": "^1.4.1",
    "@types/cors": "^2.8.5",
    "@types/dotenv-flow": "^0.2.1",
    "@types/express": "^4.17.0",
    "@types/express-session": "^1.15.13",
    "@types/hapi__joi": "^15.0.3",
    "@types/helmet": "^0.0.43",
    "@types/jest": "^24.0.11",
    "@types/lodash": "^4.14.136",
    "@types/morgan": "^1.7.36",
    "@types/passport": "^1.0.0",
    "@types/passport-facebook-token": "^0.4.33",
    "@types/passport-jwt": "^3.0.1",
    "@types/passport-local": "^1.0.33",
    "@types/redis": "^2.8.13",
    "@types/request-promise-native": "^1.0.16",
    "@types/shelljs": "^0.8.5",
    "@types/stripe": "^6.31.4",
    "@types/twilio": "^0.0.10",
    "@typescript-eslint/eslint-plugin": "^1.5.0",
    "@typescript-eslint/parser": "^1.5.0",
    "eslint": "^5.16.0",
    "eslint-config-prettier": "^4.1.0",
    "eslint-plugin-prettier": "^3.0.1",
    "jest": "^24.5.0",
    "jest-date-mock": "^1.0.7",
    "prettier": "^1.16.4",
    "rimraf": "^2.6.3",
    "shelljs": "^0.8.3",
    "ts-jest": "^24.0.1",
    "ts-node": "^8.3.0",
    "ts-node-dev": "^1.0.0-pre.40",
    "typescript": "^3.4.1"
  }
}
```

`tsconfig.json`:
```json
{
  "compilerOptions": {
    "module": "commonjs",
    "esModuleInterop": true,
    "target": "es6",
    "noImplicitAny": true,
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "strict": true,
    "noUnusedLocals": true,
    "moduleResolution": "node",
    "sourceMap": true,
    "outDir": "dist",
    "baseUrl": ".",
    "paths": {
      "*": ["node_modules/*", "src/types/*"]
    },
    "declaration": true
  },
  "include": ["src/**/*"]
}
```

Error message is there either with TS 3.4.x or TS 3.5.x.
ESLint highlights the express route declaration, and `tsc` emits the error in the console, exactly the same one as the topicstarter had.

**Comment by evanshortiss** [comment]
@hijarian can you provide the output from `./node_modules/.bin/tsc -v`?

Will try recreate with your settings.

**Comment by evanshortiss** [comment]
@hijarian confirmed the issue. Seems to be linked to `"strict": true` - I usually use `strict` too, but didn't with this module so will need to investigate a fix.

**Comment by evanshortiss** [comment]
@mxs42 @hijarian when using strict mode you'll need to change the request handler from this:

```ts
(req: ValidatedRequest<HelloRequestSchema>, res) => {
  res.end(`Hello ${req.query.name}`)
}
```

to this:

```ts
(req, res) => {
  const vreq = req as ValidatedRequest<HelloRequestSchema>
  res.end(`Hello ${vreq.query.name}`)
}
```

I'll need to think about solutions. Maybe the module needs two interfaces, one that includes `originalBody` etc. and one that does not?

```ts

import { ContainerTypes, createValidator, ValidatedRequest, ValidatedRequestWithOriginalFields } from "express-joi-validation";

const withOriginals = (req, res) => {
  const vreq = req as ValidatedRequestWithOriginalFields<HelloRequestSchema>
  const originalBody = req.originalBody
  res.end(`Hello ${req.query.name}`)
}

const withoutOriginals = (req: ValidatedRequest<HelloRequestSchema>, res) => {
  // const originalBody = req.originalBody 
  // originalBody not available with ValidatedRequest, must use ValidatedRequestWithOriginalFields


  res.end(`Hello ${req.query.name}`)
}
```

Another alternative would be to create a `req.expressJoiValidation` and assume it's globally available on all request objects (this requires the typedefs for express-joi-validation extending the express types). Then you can use this without casting:

```ts
const withoutOriginals = (req: ValidatedRequest<HelloRequestSchema>, res) => {
  const originalBody = req.expressJoiValidation.originalBody
  res.end(`Hello ${req.query.name}`)
}
```

**Comment by evanshortiss** [comment]
@hijarian @mxs42 any thoughts on suggested solutions? I prefer the latter - removing `originalBody` etc from `ValidatedRequest` and adding a `req.ejv.originalBody`, `req.ejv.originalQuery` etc.

**Comment by hijarian** [comment]
@evanshortiss I prefer the second solution, with two separate interfaces.

- It expresses intent the best
- if you actually want the original untouched body after applying Joi you can afford having to explicitly typecast
- and, most importantly, current code which works without `"strict": true` will just continue to work.

**Comment by hijarian** [comment]
> @hijarian can you provide the output from `./node_modules/.bin/tsc -v`?
> 
> Will try recreate with your settings.

This command prints just one line `Version 3.5.3` and that's it.

**Comment by mxs42** [comment]
I was on vacation so excuse me for ignoring this issue.
I prefer second solution as well, much cleaner.

**Comment by evanshortiss** [comment]
Great, that's 3/3. Let's go with option 2 👍 

Will not add `req.ejv`. It can be reconsidered at another point in if necessary.

Edit: I hope you had a great vacation @mxs42, thanks for the response.

**Comment by mxs42** [comment]
@evanshortiss indeed, thank you!

**Comment by evanshortiss** [comment]
@mxs42 @hijarian I opened #21, will merge in the coming days if you think it looks ok.

**Comment by evanshortiss** [comment]
Fixed in 3.0.0. Thanks for the reports 👍
